classes/ForbiddenSeal.php

```diff
@@ -1,336 +1,398 @@
+        public int $free_transfer_bonus,
+        public float $long_training_time,
+        public float $long_training_gains,
+        public float $extended_training_time,
+        public float $extended_training_gains,
+
+        // Misc. benefits
+        public int $bonus_pve_reputation,
+        public int $extra_jutsu_equips,
+        public int $extra_armor_equips,
+        public int $extra_weapon_equips,
+        public int $max_battle_history_view,
+
+        // Ancient Market displays
+        public string $name_color_display,
+        public string $avatar_size_display,
+        public string $journal_image_display
+    ){}
 
-    public function checkExpiration() {
-        // Seal expired, remove the seal
+    public function checkExpiration(): void {
         if($this->level > 0 && time() > $this->seal_end_time) {
-            $this->system->message("Your " . self::$forbidden_seal_names[$this->level] . " has expired!");
-            $this->level = 0;
-            $this->setBenefits();
+            // Todo: This is for display purposes only. Make this a notification
+            $this->system->message("Your {$this->name} has expired!");
         }
     }
-
-    public function addSeal(int $seal_level, int $days_to_add) {
-        //Add time
+    public function addSeal(int $seal_level, int $days_to_add): void {
+        // Add time
         if($seal_level == $this->level) {
-            $this->seal_end_time += $days_to_add * self::$SECONDS_IN_DAY;
-            $this->seal_time_remaining += $days_to_add * self::$SECONDS_IN_DAY;
+            $this->seal_end_time += $days_to_add * self::SECONDS_IN_DAY;
+            $this->seal_time_remaining += $days_to_add * self::SECONDS_IN_DAY;
         }
-        //Overwrite seal
+        // Overwrite seal
         else {
             $this->level = $seal_level;
-            $this->name = self::$forbidden_seal_names[$this->level];
-            $this->seal_end_time = time() + ($days_to_add * self::$SECONDS_IN_DAY);
-            $this->seal_time_remaining = $this->seal_end_time - time();
+            $this->name = self::$forbidden_seal_names[$seal_level];
+            $this->seal_end_time = time() + ($days_to_add * self::SECONDS_IN_DAY);
+            $this->seal_time_remaining = $days_to_add * self::SECONDS_IN_DAY;
         }
     }
-
-    public function dbEncode() {
-        switch($this->level) {
-            case 0:
-                return "";
-            default:
-                return json_encode(array('level' => $this->level, 'time' => $this->seal_end_time));
+    public function dbEncode(): string {
+        return match($this->level) {
+            0 => "",
+            default => json_encode(array('level' => $this->level, 'time' => $this->seal_end_time))
+        };
+    }
+    public function calcRemainingCredit(): int {
+        match($this->level) {
+            1 => $days_per_ak = 6,
+            2 => $days_per_ak = 2,
+            default => $days_per_ak = 1
```

Comment:
Ideally this is something we calculate from the place we define the costs of each tier, as is if we changed the costs in the future this wouldn't automatically be synchronized and there isn't necessarily a way for us to remember we need to change this